### PR TITLE
Authenticator with multiple aliases

### PR DESCRIPTION
**Pull request by slaymance**
*Issue #, if available:* [Asana](https://app.asana.com/0/1200705313127344/1200546485454778)

*Description of changes:* This PR adds the environment, example application, and end-to-end tests to capture signing in with multiple login mechanisms (in this case username, email, or phone number) as configured using Amplify CLI.

Configuring a backend environment is currently only possible using the Amplify CLI (refer to the README for instructions). Configuring an environment in the Amplify Admin UI to allow for email AND phone number authentication doesn't currently function properly.

Furthermore, by configuring an environment to allow authentication using email and/or phone through the Amplify CLI, username is always a required attribute and treated as a valid login mechanism (even if not selected as a login mechanism in the CLI prompt).

### Create a user in the `auth-with-multi-alias` environment

1. Navigate to `User management` in the Admin UI (use the environment `authwithmultialias`)
2. Click `Create user`
3. Set a unique username
4. Set the contact option for verification as "Email address" or "Phone number" and fill out a unique value for the field
5. Fill out the temporary password and click `Create user`
6. Find the user you just create and click its login alias to access its profile
7. Click `Edit` and fill out the secondary login mechanism (this alias will not work when attempting to login)
8. Check both aliases as `Verified` then click `Save`

https://user-images.githubusercontent.com/26472139/130111724-50f61f6e-64ff-48e9-963c-eebe3d8e0734.mp4

### Attempt to login

1. Pull down this branch and start the example application with `yarn next-example dev`
2. Navigate to `http://localhost:3000/ui/components/authenticator/auth-with-multi-alias`
3. Attempt to login with the three aliases you created for your user using the same temporary password you used:
  - Unique username
  - Email address
  - Phone number
4. You should be taken to a page to change your password for each alias, indicating that authenticating with that alias works

https://user-images.githubusercontent.com/26472139/130099568-b26b8d9e-e6b1-4184-a0db-ed61bcd5351e.mp4

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by eddiekeller** [comment]
```
Furthermore, by configuring an environment to allow authentication using email and/or phone through the Amplify CLI, username is always a required attribute and treated as a valid login mechanism (even if not selected as a login mechanism in the CLI prompt).
```

Lmao. So it is impossible to just have `email || phone`? It would always be `username && (email || phone)`?

**Review by eddiekeller — APPROVED** [review]

File: environments/auth-with-multi-alias/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+1. In your application's root directory, add authentication using the Amplify CLI:
+
+   ```shell
+   amplify add auth
+   ```
+
+1. When prompted `Do you want to use the default authentication and security configuration?`, select `Manual configuration`
+1. When prompted `How do you want users to be able to sign in?`, select the aliases you would like users to be able to use to sign in to your application
+   ![](multi-alias-config.png)
```
Comment (ericclemmons) [inline environments/auth-with-multi-alias/README.md:17]: The image is kinda inline here https://github.com/aws-amplify/amplify-ui/blob/26667e59f49853772c6db94e67cd3590acf68bfb/environments/auth-with-multi-alias/README.md, so let's give it some breathing room:

```suggestion
1. When prompted `How do you want users to be able to sign in?`, select the aliases you would like users to be able to use to sign in to your application:

    > ![](multi-alias-config.png)
```

File: environments/auth-with-multi-alias/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+   _Note: a username is still required when a user creates an account and is treated as a sign in alias regardless of if it was selected as an option._
```
Comment (ericclemmons) [inline environments/auth-with-multi-alias/README.md:19]: In your image, only `username` and `phone_number` are selected.

Does that mean that the `username` value === `email`?  Or that, just like your demo, the backend will return `['username', 'email', 'phone_number']`?

File: examples/next/pages/ui/components/authenticator/auth-with-multi-alias/index.page.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { Authenticator } from '@aws-amplify/ui-react';
+import { Amplify } from 'aws-amplify';
+import awsExports from '@environments/auth-with-multi-alias/src/aws-exports';
+
+Amplify.configure({
+  ...awsExports,
+  auth: {
+    login_mechanisms: ['username', 'email', 'phone_number'],
```
Comment (ericclemmons) [inline examples/next/pages/ui/components/authenticator/auth-with-multi-alias/index.page.tsx:8]: We really need that `amplify pull` update to land so we can see what _actually_ comes back.

**Review by ericclemmons — APPROVED** [review]
Despite the `username` weirdness, this solves a long-standing UX issue.

Great work @slaymance

File: environments/auth-with-multi-alias/README.md (hunk 2 above)
Comment (slaymance) [reply environments/auth-with-multi-alias/README.md:19]: Even without `username` selected, it's still treated as a valid login mechanism and is required for creating an account. It's a bit strange, but one may think of it as `username` always being a selected option.

File: examples/next/pages/ui/components/authenticator/auth-with-multi-alias/index.page.tsx (hunk 3 above)
Comment (slaymance) [reply examples/next/pages/ui/components/authenticator/auth-with-multi-alias/index.page.tsx:8]: Definitely. I take some comfort in the fact that our UI will automatically handle whatever login mechanisms are returned. We'll just need to make sure that the login mechanisms consumed as part of zero-config match what's actually possible in Cognito. I feel it best we provide guide rails for customers as much as possible.

**Pull request merged**